subsys/net/lib/download_client/src/coap.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <net/coap.h>
+#include <net/download_client.h>
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(download_client, CONFIG_DOWNLOAD_CLIENT_LOG_LEVEL);
+
+#define COAP_VER 1
+
+extern int socket_send(const struct download_client *client, size_t len);
+
+int coap_block_init(struct download_client *client, size_t from)
+{
+	coap_block_transfer_init(&client->coap.block_ctx,
+				 CONFIG_DOWNLOAD_CLIENT_COAP_BLOCK_SIZE, 0);
+	client->coap.block_ctx.current = from;
+	return 0;
+}
+
+int coap_block_update(struct download_client *client, struct coap_packet *pkt,
+		      size_t *blk_off)
+{
+	int err;
+	bool more;
+
+	*blk_off = client->coap.block_ctx.current %
```

Comment:
I'm a bit lost here, what's the scenario when we receive only part of the block?

Reply:
So what's happening here is that we request to start from an offset (in `_start(..., from)`) that is not aligned to the block size. What the server does, is to re-send that block that contains those bytes and we don't want to process those again.

In detail we set `block_ctx.current` to `from` when we initialize the download, say for example 100 with block size 64.
We'll request block number 2 and the server will send bytes from 64 to 128, of these we want to discard `blk_off`, `100 % 64 = 36` and copy what's left, so in `coap_parse` where we do the copy we copy from `payload + blk_off` , a total of `payload_len - blk_off` bytes.

Reply:
Thanks for exaplaining